powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ActivationServiceBehavior.java

```diff
@@ -1701,52 +1715,61 @@ public UpdateActivationOtpResponse updateActivationOtp(UpdateActivationOtpReques
-            // This should never happen.
-            logger.info("Internal error in activation OTP validation: {}", activationId);
-            // Rollback is not required, database is not used for writing yet.
-            throw localizationProvider.buildExceptionForCode(ServiceError.UNKNOWN_ERROR);
-        }
-
-        // Check whether activation OTP validation is turned OFF. In this case, the confirmation OTP must not
-        // be provided.
-        if (expectedStage == com.wultra.security.powerauth.client.model.enumeration.ActivationOtpValidation.NONE) {
-            if (confirmationOtp != null) {
-                logger.info("Activation OTP is not used, but is provided: {}", activationId);
+        if (expectedStage != ActivationOtpValidation.NONE) {
+            // Validation using ActivationOtpValidation (deprecated)
+            // Validate activation OTP is present for given stage, if required.
+            if (expectedStage == currentStage && !StringUtils.hasText(confirmationOtp)) {
+                logger.info("Activation OTP is missing for activation ID: {}, stage: {}", activationId, currentStage);
                 // Rollback is not required, database is not used for writing yet.
                 throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
             }
-            return;
-        }
-        // Check whether this is validation in the different step of activation. If yes, then the confirmation
-        // OTP must not be provided.
-        if (expectedStage != currentStage) {
-            if (confirmationOtp != null) {
-                logger.info("Activation OTP is not expected, but is provided: {}", activationId);
+            // Validate activation OTP is absent when it should not be present.
+            if (expectedStage != currentStage) {
+                if (StringUtils.hasText(confirmationOtp)) {
+                    logger.info("Activation OTP should not be present for activation ID: {}, stage: {}", activationId, currentStage);
+                    // Rollback is not required, database is not used for writing yet.
+                    throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
+                }
+                // No validation required for this stage, the current stage is different from expected stage
+                return;
+            }
+        } else {
+            // Validation of commit phase:
+            // - if no OTP is specified for an activation, no validation is required
+            // - check each stage to make sure OTP is available when required
+            // - in case the commit is done in the next stage, skip validation
+            if (expectedOtpHash == null) {
+                return;
+            }
+            if (currentStage == ActivationOtpValidation.ON_KEY_EXCHANGE && commitPhase == CommitPhase.ON_KEY_EXCHANGE && !StringUtils.hasText(confirmationOtp)) {
```

Comment:
Like the description. I suggest:
- Change currentStage from ActivationOtpValidation to CommitPhase enum (this is to prepare for the future when activationOtpValidation is going to be dropped).
- Change the name of variable `commitPhase` to `expectedCommitPhase` (to distinguish what is expected by activation setting and what is the phase of the control)
This will increase the clarity of the code.

Reply:
Good idea.